File: mobile/ios/RaverMVP/RaverMVP/Features/Search/Storage/RecentSearchStore.swift
import Foundation

@MainActor
final class RecentSearchStore: ObservableObject {
    @Published private(set) var queries: [String] = []

    private let defaults: UserDefaults
    private let key: String
    private let maxCount: Int

    init(
        defaults: UserDefaults = .standard,
        key: String = "globalSearch.recentQueries.v1",
        maxCount: Int = 10
    ) {
        self.defaults = defaults
        self.key = key
        self.maxCount = max(1, maxCount)
        self.queries = Self.normalizedList(defaults.stringArray(forKey: key) ?? [], maxCount: max(1, maxCount))
    }

    func record(_ rawQuery: String) {
        let query = rawQuery.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !query.isEmpty else { return }

        var next = queries.filter { $0.caseInsensitiveCompare(query) != .orderedSame }
        next.insert(query, at: 0)
        queries = Self.normalizedList(next, maxCount: maxCount)
        defaults.set(queries, forKey: key)
    }

    func clear() {
        queries = []
        defaults.removeObject(forKey: key)
    }

    private static func normalizedList(_ items: [String], maxCount: Int) -> [String] {
        var seen: Set<String> = []
        var result: [String] = []
        for item in items {
            let trimmed = item.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty else { continue }
            let key = trimmed.lowercased()
            guard !seen.contains(key) else { continue }
            seen.insert(key)
            result.append(trimmed)
            if result.count >= maxCount {
                break
            }
        }
        return result
    }
}

File: mobile/flutter_ravehub/packages/features/feature_discover/lib/src/search/data/recent_search_store.dart
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

class RecentSearchStore extends ChangeNotifier {
  RecentSearchStore({
    this.key = 'globalSearch.recentQueries.v1',
    this.maxCount = 10,
  });

  final String key;
  final int maxCount;

  List<String> _queries = [];
  List<String> get queries => List.unmodifiable(_queries);

  bool _initialized = false;

  Future<void> initialize() async {
    if (_initialized) return;
    final prefs = await SharedPreferences.getInstance();
    final saved = prefs.getStringList(key) ?? [];
    _queries = _normalize(saved);
    _initialized = true;
    notifyListeners();
  }

  Future<void> record(String rawQuery) async {
    final query = rawQuery.trim();
    if (query.isEmpty) return;

    final next = _queries
        .where((q) => q.toLowerCase() != query.toLowerCase())
        .toList();
    next.insert(0, query);
    _queries = _normalize(next);
    notifyListeners();

    final prefs = await SharedPreferences.getInstance();
    await prefs.setStringList(key, _queries);
  }

  Future<void> clear() async {
    _queries = [];
    notifyListeners();

    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(key);
  }

  List<String> _normalize(List<String> items) {
    final seen = <String>{};
    final result = <String>[];
    for (final item in items) {
      final trimmed = item.trim();
      if (trimmed.isEmpty) continue;
      final lowerKey = trimmed.toLowerCase();
      if (seen.contains(lowerKey)) continue;
      seen.add(lowerKey);
      result.add(trimmed);
      if (result.length >= maxCount) break;
    }
    return result;
  }
}
